### PR TITLE
feat: add support for plane trips

### PR DESCRIPTION
**Pull request by NTag**
# howmuch.green/Paris+new+york

This PR adds a base and a first support for **trips carbon emissions estimation**.

CleanShot 2022-07-07 at 00 21 38@2x

## How does it work?
From a query string like `paris new york`, `sao polo san francisco`, `helsinki auckland` it extracts the two most probable cities, using the levenshtein distance against the list of world cities with more than 10k people (from the npm package [all-the-cities](https://www.npmjs.com/package/all-the-cities)). Then it simply computes the distance between the two cities and apply a coefficient (currently 158g of CO2eq per km, but a lot of sources vary between 100g and 200g).

Actually it's surprisingly fast 🔥 .

## Other changes
Because the list of the biggest world cities (with their location and countries) weights several MB, it wasn't wise to include it when the library is used in the browser. So the library has now two different bundles:
- one for NodeJS, which depends on [all-the-cities](https://www.npmjs.com/package/all-the-cities), which includes the file `things.json` and the logic to find things or trips (basically everything);
- one for the browser, which includes only the domain part. This one is much smaller and allows to manipulate the objects received serialized from a backend.

I struggled doing it with rollup (also what is this doc they have 🤮), so I switched to Parcel. It wasn't that easy either, but I finally managed to do it.

## Notes
I chose a typing that could evolve in the future: a Trip has just an origin/destination, and contains Transports, which for now has always one entry (flying) but that could be enriched tomorrow with a train/bike/car alternative, so people could see the different options in the website and compare them together.

Instead of presenting the general function `howMuch()` in the Readme, it presents two functions, so users can limit the search specifically to a thing or to a trip (because I think use-cases for `howMuch()` would be quit rare appart howmuch.green).

Also I'm not really sure about the code in the domain and fp-ts/io-ts part. I wonder if the whole thing couldn't be simplified, it feels like a lot of boilerplate to me sometimes 😬 . But that could be the point of other discussions and PRs, I'm happy I've been able to make it work from end to end 🙌 .

Of course, totally open to changes about anything.

### CODE REVIEW

File: library/src/howMuch.ts [hunk 1]
```diff
@@ -1,20 +1,20 @@
...
+export const howMuch = (query: string): ResultObject => {
+  const results: (Trip | Thing)[] = [];
 
-if (isLeft(things)) {
-  throw new Error(
-    `There was a problem decoding thing data: ${draw(things.left)}`
-  );
-}
+  const thingsResults = getThingsFromQuery(query);
+  const tripResult = getTripFromQuery(query);
```
Comment (licarth) [inline library/src/howMuch.ts:9]: Not a fan of APIs like getXFrom or getXBy ... 
Since we have Typescript, why not shorter names and more explicit types ?
From a vocabulary standpoint, `getFromQuery` = `search`.
So why not `searchThings(ThingQueryString)` and `searchTrips(TripQueryString)`

If we want to overload `searchThings` or `searchTrips` in the future then we couold even keep it shorter like `things(ThingQueryString)` and `trips(TripQueryString)`.

File: library/README.md [hunk 2]
```diff
@@ -15,18 +18,27 @@ npm i @howmuchgreen/howmuchcarbon
...
+} from "@howmuchgreen/howmuchcarbon";
 
-const macbookPro14 = howMuch("macbook pro 14").bestResult;
+const [macbookPro14] = getThingsFromQuery("macbook pro 14");
 console.log(macbookPro14.name); // 'MacBook Pro 14' 2021'
 console.log(macbookPro14.co2Eq.averageInGrams); // 271000
 console.log(macbookPro14.co2Eq.format()); // '271 kg'
 console.log(macbookPro14.sources); // ['https://www.apple.com/environment/…']
+
+const tripParisNewYork = getTripFromQuery("paris new york")?.trip!;
```
Comment (licarth) [inline library/README.md:32]: Why could the trip be `null` or `undefined` ? `getTripFromQuery()` returns some result, I expect the `trip` property to be non-null, nope ?

File: library/src/trips/index.ts [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+  const distanceInKm = getDistanceFromLatLonInKm(
+    origin.loc.coordinates,
+    destination.loc.coordinates
+  );
+
+  const carbonInKg = distanceInKm * CO2EQ_KG_PER_KM_FLIGHT;
+
+  const trip = Trip.build({
+    kind: "trip",
```
Comment (licarth) [inline library/src/trips/index.ts:138]: This can be done in the `build()` method

File: library/src/trips/index.ts [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+      country: origin.country,
+    },
+    destination: {
+      name: destination.name,
+      country: destination.country,
+    },
+    distanceInKm: distanceInKm,
+    transports: [
+      {
+        type: "flight",
```
Comment (licarth) [inline library/src/trips/index.ts:150]: use `Transport.build()`

File: library/src/domain/Transport.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
+import { pipe } from "fp-ts/function";
+import * as Codec from "io-ts/Codec";
+import { Co2Eq } from "./Co2Eq";
+import { fromClassCodec } from "./io-ts";
+
+export class Transport {
+  type;
+  co2Eq;
+  explanation;
```
Comment (licarth) [inline library/src/domain/Transport.ts:9]: `TransportMode` ? `TransportType` ? GTFS uses `RouteType` for this, but we don't have Routes

File: library/src/howMuch.ts (hunk 1 above)
Comment (licarth) [reply library/src/howMuch.ts:9]: And why not keep `howMuch()` on top of these 2 ?

File: library/src/trips/index.ts [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+  const d = R * c; // Distance in km
+  return d;
+};
+
+const CO2EQ_FLIGHT_SOURCES = ["https://www.carbonindependent.org/22.html"];
+const CO2EQ_KG_PER_KM_FLIGHT = 0.158;
+
+export const getTripFromQuery = (
+  query: string
+): { trip: Trip; score: number } | undefined => {
```
Comment (licarth) [inline library/src/trips/index.ts:119]: Why not have a similar signature than for Things ? 
`getTripsFromQuery(): TripWithScore[]`

Comment (licarth) [reply library/src/trips/index.ts:119]: I'm thinking about typing "Paris Br" and have results like
"Paris > Brussels"
"Paris > Brest" (which could also be a Thing 🍰 )

File: library/README.md (hunk 2 above)
Comment (licarth) [reply library/README.md:32]: Oh I see it's because of the `getTripFromQuery("paris new york")?` that it can be null or undefined :)
So why not put an exclamation mark on `getTripFromQuery("paris new york")!` ?

**Comment by licarth** [comment]
`usage-tests/cra` & `usage-tests/nodejs` are broken. howMuch() does not exist anymore + I see the following error in CRA when I try to use the library. Not sure why.

CleanShot 2022-07-16 at 10 05 45@2x

File: library/src/howMuch.ts (hunk 1 above)
Comment (NTag) [reply library/src/howMuch.ts:9]: Agreed on shorter names, I think `searchThings` and `searchTrips` are indeed very clear.

However, I wouldn't rely too much on types: whereas it's a great help when coding, then when reading code later or in a pull request, you don't see the types (except if you plan to place your cursor hover every variable 😬), so I think the names should be clear enough so you understand the code without the types.

For example, if I read something like `import { things } from 'howmuchcarbon';` I would expect `things` to be some data, like an object or an array. So I wouldn't advise we shorten that much the names. And usually starting functions with a verb helps to make it clear it's a function and not a variable.

But like I said, totally fine with `searchThings` and `searchTrips`, which are shorter and still clear, I'll do the changes 👍

Comment (NTag) [reply library/src/howMuch.ts:9]: About `howMuch()` in the README, like I said in the PR description, I don't really see use-cases appart us where users would need it, so for clarity I suggested to remove it and focus on individual functions.

But I can add it back if you want, I don't have a strong opinion here.

File: library/README.md (hunk 2 above)
Comment (NTag) [reply library/README.md:32]: Actually, like you said below, I agree that `searchTrips` should return an array `TripWithScore[]`. I had the same idea, and I thought about changing the code to return multiple trips but it wasn't that easy so I decided to open the PR with only a single trip for now; however we can still return an array (with one element for now) to prepare for the future 👍 

So this part of the readme will de facto change.

File: library/src/trips/index.ts (hunk 3 above)
Comment (NTag) [reply library/src/trips/index.ts:138]: Ok, I'll try to do that

File: library/src/trips/index.ts (hunk 4 above)
Comment (NTag) [reply library/src/trips/index.ts:150]: Ok, but why?

File: library/src/trips/index.ts (hunk 6 above)
Comment (NTag) [reply library/src/trips/index.ts:119]: Yes agreed, as said above, I'll do the change 👍

File: library/src/domain/Transport.ts (hunk 5 above)
Comment (NTag) [reply library/src/domain/Transport.ts:9]: Actually Transport may also be a combination of multiple means. For example, on the page for the search "Paris - Ajaccio", we should display two alternatives:
- the plane option
- the train + boat option

So the current structure may not be enough, new proposal:
- `Trip` still represents the wish of the user: traveling from `origin` to `destination`.
- `Trip.routes: Route[]`, a `Route` would represent a solution to go from `origin` to `destination`. It has a total co2eq. It can be composed of multiple means of transportation, so:
- `Route.segments: Segment[]`, to represent the details of a route, a `Segment` has its own `origin & destination`, a `type: flight | train | boat`, also a co2eq (`sum(route.segments, segment => segment.co2eq) === route.co2eq`), why not a distance

To simplify the front, we could add a `type` to `Route` which could be `'flight' | 'train' | 'boat' | 'train+boat'` depending on the segments.

**Comment by NTag** [comment]
`usage-tests/nodejs` seems to work, no? Don't forget to do a `npm run dev` in the library folder to build it, and a `npm install` in `usage-tests/nodejs`
CleanShot 2022-07-16 at 18 30 09@2x

Indeed there is an issue with CRA, I'm investigating 👀

**Comment by NTag** [comment]
Ah no alright, I get why CRA isn't working anymore: as explained in the description, the bundle for the browser now only includes the domain part, it doesn't include `howMuch, searchTrips, searchThings` (but as we can only have one `types` file, you think you can import `howMuch` whereas you can't).

The library destiny is to store more and more data, and get quit big, I really think it doesn't really make sense to use it in a browser context (expect the domain part). Should we remove the CRA example? Or update it to only use the domain objects? Not sure it's very useful…

And if we remove it, I'm not sure if the node example is very useful too, as it would be basically the README.

**Comment by licarth** [comment]
I disagree there :)

It's very weird to have types but not have the corresponding implementations. Can we investigate the solution where we split the data in a separate bundle and fail at runtime if data isn't there ?

Instead of removing the CRA & nodejs examples, I'd actually add those to the CI.

The things may not be very very big in the end if we properly compress data in the bundle. If someone wants to use this in an offline mobile app, they would not be able to. And I don't think it's difficult enough to do technically to be a good argument for us not to do it.

That said, I'm happy to try to spend some time on it on my side (at least to have tests in the CI for CRA & nodejs usages)

**Comment by NTag** [comment]
> Can we investigate the solution where we split the data in a separate bundle and fail at runtime if data isn't there

That wouldn't change anything on the types, would it? 🤔 

However I understand your arguments, I agree it's not ideal. I can check to have an explicit browser bundle and two types declarations. But I already spent quit some time just to make that work, it would be nice that this PR doesn't lend in one month 🥲

> I don't think it's difficult enough to do technically to be a good argument for us not to do it

On my side I think it's a very good argument to at least report that. Given that we don't have a lot of time to give to this project, prioritization is then very important. We must have a "lean approach" where we prioritize things with the most impact and yes, take into account the ratio time/benefit. Also, it's not because today we don't do or support something that it'll be like that forever ;) .

So in conclusion, I'm ok to spend some time again to dig into the types generation as I don't like either the current situation (but if too long/complex, I'd advise we start like that, at least to have the feature merged and available, and we can always improve after), but I won't spend some time right now to make the library fully work in the browser, as the benefit/use-case in the short term is very very low to me, and also because howmuch.green isn't an offline app and it's completely out of question that we load 10 MB of data: we'll have to have two bundles anyway (or we just drop the library in the front and directly read the JSON returned by the API, but it'd be shame).

**Comment by NTag** [comment]
Ok, so I pushed a new commit that computes the types for the second bundle 🙌 .

In NodeJS, you can `import { howMuch, Trip, Thing } from '@howmuchgreen/howmuchcarbon';`, nothing changes.

But in the browser, you need to import from `@howmuchgreen/howmuchcarbon/domain` and then you have a type error if you try to import `howMuch` for example:
CleanShot 2022-07-16 at 20 35 51@2x

I also pushed a new commit on the website PR.

On my side I'm happy with this solution. I think we could delay to later to idea of having the whole node bundle also work in the browser? Of course if you find an easy solution soon, no problem, but on my side I don't want to spend time now on this use-case (as I really don't see the benefits in the short term), and I'd like to have this PR merged not too far in the future 🤞.

File: library/src/trips/index.ts (hunk 3 above)
Comment (NTag) [reply library/src/trips/index.ts:138]: I thought I succeeded but no, I tried different things but either `kind` disappears when serializing the object to JSON, either I have TS issues.

If you have the solution, happy to hear it!

**Comment by licarth** [comment]
I feel that this is going to be a debate we're going to have again :)

Other libraries have similar issues :
- date libraries with locales. `date-fns` is tree-shakable and this is how users import locales:
  ```javascript
  import { formatDistance } from 'date-fns'
  // Require Esperanto locale
  import { eo } from 'date-fns/locale'
  
  const result = formatDistance(
  new Date(2016, 7, 1),
  new Date(2015, 0, 1),
  {locale: eo} // Pass the locale as an option
  )
  //=> 'pli ol 1 jaro'
  ```

  We could imagine doing something like this for us:

  ```typescript
  import { howMuch } from '@howmuchgreen/howmuchcarbon'
  import { things } from '@howmuchgreen/things/all'
  import { trips } from '@howmuchgreen/trips/biggest-cities-1000'

  console.log(new HowMuch({things, trips}).howMuch("paris berlin"))
  console.log(new HowMuch({things, trips}).searchThings("ipad"))
  ```

  ```typescript
  import { howMuch } from '@howmuchgreen/howmuchcarbon'
  import allData from '@howmuchgreen/howmuchcarbon/all-data'

  console.log(new HowMuch(allData).howMuch("lion-sur-mer paris"))
  ```

File: library/src/trips/types.d.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+    featureCode: string;
+    adminCode: string;
+    population: number;
+    loc: { type: "Point"; coordinates: LngLat };
+  };
+
+  const Cities: City[];
+
+  export default Cities;
+}
```
Comment (licarth) [inline library/src/trips/types.d.ts:18]: Why not use https://www.npmjs.com/package/@types/all-the-cities ?

Comment (NTag) [reply library/src/trips/types.d.ts]: Oups I haven't seen it, updated 👍

**Review by licarth — APPROVED** [review]
